# Explore data



## Import necessary libraries

In [62]:
import pandas as pd
import numpy as np

## Read data from file

In [63]:
# Read data from file
ds_survey_df = pd.read_csv('../Dataset/kaggle_survey_2022_responses.csv', )
ds_survey_df.head()

C:\Users\DELL\AppData\Local\Temp\ipykernel_4768\2297636855.py:2: DtypeWarning: Columns (0,15,43,57,73,88,104,118,126,132,170,200,208,215,225,248,255,257,260,270,271,272,277,281,294) have mixed types. Specify dtype option on import or set low_memory=False.
  ds_survey_df = pd.read_csv('../Dataset/kaggle_survey_2022_responses.csv', )


Duration (in seconds)                           Q2  \
0  Duration (in seconds)  What is your age (# years)?   
1                    121                        30-34   
2                    462                        30-34   
3                    293                        18-21   
4                    851                        55-59   

                                       Q3  \
0  What is your gender? - Selected Choice   
1                                     Man   
2                                     Man   
3                                     Man   
4                                     Man   

                                          Q4  \
0  In which country do you currently reside?   
1                                      India   
2                                    Algeria   
3                                      Egypt   
4                                     France   

                                                  Q5  \
0  Are you currently a student? (high school, uni...   
1                                                 No   
2                                                 No   
3                                                Yes   
4                                                 No   

                                                Q6_1  \
0  On which platforms have you begun or completed...   
1                                                NaN   
2                                                NaN   
3                                           Coursera   
4                                           Coursera   

                                                Q6_2  \
0  On which platforms have you begun or completed...   
1                                                NaN   
2                                                NaN   
3                                                edX   
4                                                NaN   

                                                Q6_3  \
0  On which platforms have you begun or completed...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                               Kaggle Learn Courses   

                                                Q6_4  \
0  On which platforms have you begun or completed...   
1                                                NaN   
2                                                NaN   
3                                           DataCamp   
4                                                NaN   

                                                Q6_5  ...  \
0  On which platforms have you begun or completed...  ...   
1                                                NaN  ...   
2                                                NaN  ...   
3                                                NaN  ...   
4                                                NaN  ...   

                                               Q44_3  \
0  Who/what are your favorite media sources that ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                               Q44_4  \
0  Who/what are your favorite media sources that ...   
1                                                NaN   
2                                                NaN   
3                    Kaggle (notebooks, forums, etc)   
4                    Kaggle (notebooks, forums, etc)   

                                               Q44_5  \
0  Who/what are your favorite media sources that ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  Course Forums (forums.fast.ai, Coursera forums...   

                                     

In [64]:
ds_survey_df.drop(0,inplace = True) 
ds_survey_df.reset_index(drop= True,inplace = True)
ds_survey_df.head()


Duration (in seconds)     Q2   Q3       Q4   Q5      Q6_1 Q6_2  \
0                   121  30-34  Man    India   No       NaN  NaN   
1                   462  30-34  Man  Algeria   No       NaN  NaN   
2                   293  18-21  Man    Egypt  Yes  Coursera  edX   
3                   851  55-59  Man   France   No  Coursera  NaN   
4                   232  45-49  Man    India  Yes       NaN  NaN   

                   Q6_3      Q6_4 Q6_5  ... Q44_3  \
0                   NaN       NaN  NaN  ...   NaN   
1                   NaN       NaN  NaN  ...   NaN   
2                   NaN  DataCamp  NaN  ...   NaN   
3  Kaggle Learn Courses       NaN  NaN  ...   NaN   
4                   NaN       NaN  NaN  ...   NaN   

                             Q44_4  \
0                              NaN   
1                              NaN   
2  Kaggle (notebooks, forums, etc)   
3  Kaggle (notebooks, forums, etc)   
4                              NaN   

                                               Q44_5  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  Course Forums (forums.fast.ai, Coursera forums...   
4                                                NaN   

                                               Q44_6  \
0                                                NaN   
1                                                NaN   
2  YouTube (Kaggle YouTube, Cloud AI Adventures, ...   
3                                                NaN   
4                                                NaN   

                                               Q44_7  \
0                                                NaN   
1                                                NaN   
2  Podcasts (Chai Time Data Science, O’Reilly Dat...   
3                                                NaN   
4                                                NaN   

                                               Q44_8 Q44_9 Q44_10 Q44_11  \
0                                                NaN   NaN    NaN    NaN   
1                                                NaN   NaN    NaN    NaN   
2                                                NaN   NaN    NaN    NaN   
3  Blogs (Towards Data Science, Analytics Vidhya,...   NaN    NaN    NaN   
4  Blogs (Towards Data Science, Analytics Vidhya,...   NaN    NaN    NaN   

  Q44_12  
0    NaN  
1    NaN  
2    NaN  
3    NaN  
4    NaN  

[5 rows x 296 columns]

First, let's see how many columns and the meaning of each columns in the datasets.

In [65]:
num_rows, num_cols = ds_survey_df.shape

print("Number of columns: ", num_cols)
print("Number of rows: ", num_rows)
print("Columns: ", ds_survey_df.columns)

Number of columns:  296
Number of rows:  23997
Columns:  Index(['Duration (in seconds)', 'Q2', 'Q3', 'Q4', 'Q5', 'Q6_1', 'Q6_2', 'Q6_3',
       'Q6_4', 'Q6_5',
       ...
       'Q44_3', 'Q44_4', 'Q44_5', 'Q44_6', 'Q44_7', 'Q44_8', 'Q44_9', 'Q44_10',
       'Q44_11', 'Q44_12'],
      dtype='object', length=296)


*Notice that all columns has data types of object. This makes sense, since all the questions answer are of type string. As for Duration, the suitable data type should be float.*

---
The details meaning of the columns are aviailable as Kaggle had published them along with the survey information. 

*Note: "Person who took the survey" wiil be shortened to PWTS*

## Details meaning of each questions/columns sorted by theme:

- Durations: The duration the survey takes to complete by a person. (Calculated by system)

### Bibography:

- Q2: Age of PWTS (One choice)
- Q3: Gender of PWTS (One choice)
- Q4: Country that PWTS comes from (Once choice)
- Q5: Whether PWTS a student (One choice)

### Education:

- Q6: Platforms that PWTS have used to study Data Sciecne (Multiple choice)
- Q7: Platforms/Prodects that PWTS found to be most helpful when started studying DS (Multiple choice)
- Q8: Highest level of education that PWTS had attained/plan to attain in the next 2 years (One choice)
- Q9: Whether PWTS had published an academic research (One choice)
- Q10: If PWTS had published an academic research, did the research make use of machine learning? (Multiple choice)

### Programming skills & experience:

- Q11: Number of years the PWTS have been writing code (One choice)
- Q12: Programming languages that PWTS use on a regular basis? (Multiple choice)
- Q13: IDE that PWTS use on a regular basis? (Multiple choice)
- Q14: Notebook hosting products that PWTS use? (Multiple choice)
- Q15: Data visualization libraries that PWTS use on a regular basis? (Multiple choice)
- Q16: Number of years PWTS have used ML methods? (One choice)
- Q17: ML frameworks that PWTS use on a regular basis? (Multiple choice)
- Q18: ML algorithm that PWTS use on a regular basis? (Multiple choice)
- Q19: Computer vision methods that PWTS use on a regular basis? (Multiple choice)
- Q20: NLP (natural language processing) methods that PWTS use on a regular basis? (Multiple choice)
- Q21: Pre-trained model weights services that PWTS download? (Multiple choice)
- Q22: ML model hubs/repositories that PWTS use often? (Multiple choice)

### Employments:

- Q23: Tittle that most similar to PWTS current role? (Multiple choice)
- Q24: Industry that PWTS currently employed/contracted in? (One choice)
- Q25: Size of company that PWTS employed in? (One choice)
- Q26: The approximate number of individuals responsible for DS workloads at PWTS place of work? (One choice)
- Q27: Whether current employer incorporate ML methods in their business? (One choice)
- Q28: Activities that makes up important part of PWTS at work? (Multiple choice)
- Q29: Current yearly compensation of PWTS? (One choice)

### Other product used (Cloud computing, Data storage...):

- Q30: The approximate number of money that PWTS had spent of ML/cloud computer at home/at work in the past 5 years in USD? (One choice)
- Q31: Cloud computing platforms that PWTS uses? (Multiple choice)
- Q32: Cloud platforms that PWTS had most familiar with/has the best developer experience? (One choice)
- Q33: Cloud computing products that PWTS uses? (Multiple choice)
- Q34: Data storage products that PWTS uses? (Multiple choice)
- Q35: Data products (relational databases, data warehouses, data lakes...) that PWTS are most farmiliar with? (Multiple choice)
- Q36: Business Intelligence tools that PWTS uses? (Multiple choice)
- Q37: Managed machine learning products that PWTS uses? (Multiple choice)
- Q38: Automated Machine Learning tools that PWTS uses? (Multiple choice)
- Q39: Products that PWTS to help with ML models? (Multiple choice)
- Q40: Tools that help PWTS monitor ML models/experiments? (Multiple choice)
- Q41: Responsible/Ethical AI products that PWTS uses in ML practices? (Multiple choice)
- Q42: Specilized hardware that PWTS uses when training ML models? (Multiple choice)
- Q43: Times PWTS used TPU (tensor processing unit)? (One choice)

### Source of information:

- Q44: Favorite DS media sources of PWTS? (Multiple choice)

In total, the survey have 44 questions with number of restrictions related to the choice in some questions. For example, if the answer to Q5 answer is that PWTS's a student, questions related to employment would not be asked.

The reason the number of columns in the dataset (296) is not equal to the number of questions in the survey is that each columns represent an option/a choice in every questions.

### Rename the columns:

Currently, we can see that the name of columns is so long and it contain many extra words. Therefore, we would like to shorten them.    
 The columns can be grouped into two categories: single-choice questions and multiple-choice questions. We will rename each type of column sequentially.

In [66]:
one_choice_cols = {
    'Q2': 'Age',
    'Q3': 'Gender',
    'Q4': 'Residential country',
    'Q5': 'Student?', 
    'Q8': 'Higher education?',
    'Q9': 'Academic research publication?',
    'Q11': 'Programming experience',
    'Q16': 'ML experience',
    'Q24': 'Current industry',
    'Q25' : 'Company size',
    'Q26': 'Individual responible',
    'Q27': 'Incorporte ML methods',
    'Q29': 'Current income',
    'Q30' : 'Money spend for ML/Cloud',
    'Q32': 'Familiar Cloud Platform',
    'Q43' : 'TPU experience'
}


In [67]:

# With multiple choices columns
Q6 = {
    'Q6_1': 'DS Platform (Coursera)',
    'Q6_2': 'DS Platform (edX)',
    'Q6_3': 'DS Platform (Kaggle Learn Courses)',
    'Q6_4': 'DS Platform (DataCamp)',
    'Q6_5': 'DS Platform (Fast.ai)',
    'Q6_6': 'DS Platform (Udacity)',
    'Q6_7': 'DS Platform (Udemy)',
    'Q6_8': 'DS Platform (LinkedIn Learning)',
    'Q6_9': 'DS Platform (Cloud-certification programs)',
    'Q6_10': 'DS Platform (University Courses)',
    'Q6_11': 'DS Platform (None)',
    'Q6_12': 'DS Platform (Other)'
}

Q7 = {
    'Q7_1': 'Most helpfull Platform (University courses)',
    'Q7_2': 'Most helpfull Platform (Online courses)',
    'Q7_3': 'Most helpfull Platform (Social media platforms)',
    'Q7_4': 'Most helpfull Platform (Video platforms)',
    'Q7_5': 'Most helpfull Platform (Kaggle)',
    'Q7_6': 'Most helpfull Platform (None / I do not study data science)',
    'Q7_7': 'Most helpfull Platform (Other)',
}

Q10 = {
    'Q10_1': 'Use ML (Theoretical research)',
    'Q10_2': 'Use ML (Applied research)',
    'Q10_3': 'Use ML (No)',    
}

Q12 = {
    'Q12_1': 'Programming language (Python)',
    'Q12_2': 'Programming language (R)',
    'Q12_3': 'Programming language (SQL)',
    'Q12_4': 'Programming language (C)',
    'Q12_5': 'Programming language (C#)',
    'Q12_6': 'Programming language (C++)',
    'Q12_7': 'Programming language (Java)',
    'Q12_8': 'Programming language (JavaScript)',
    'Q12_9': 'Programming language (Bash)',
    'Q12_10': 'Programming language (PHP)',
    'Q12_11': 'Programming language (MATLAB)',
    'Q12_12': 'Programming language (Julia)',
    'Q12_13': 'Programming language (Go)',
    'Q12_14': 'Programming language (None)',
    'Q12_15': 'Programming language (Other)',
    
}

Q13 = {
    'Q13_1': 'IDE (JupyterLab)',
    'Q13_2': 'IDE (RStudio/Posit)',
    'Q13_3': 'IDE (Visual Studio)',
    'Q13_4': 'IDE (Visual Studio Code)',
    'Q13_5': 'IDE (PyCharm)',
    'Q13_6': 'IDE (Spyder)',
    'Q13_7': 'IDE (Notepad++)',
    'Q13_8': 'IDE (Sublime Text)',
    'Q13_9': 'IDE (Vim, Emacs)',
    'Q13_10': 'IDE (MATLAB)',
    'Q13_11': 'IDE (Jupyter  Notebook)',
    'Q13_12': 'IDE (IntelliJ)',
    'Q13_13': 'IDE (None)',
    'Q13_14': 'Other',
}

Q14 = {
    'Q14_1': 'Hosted notebook product (Kaggle Notebooks)',
    'Q14_2': 'Hosted notebook product (Colab Notebooks)',
    'Q14_3': 'Hosted notebook product (Azure Notebooks)',
    'Q14_4': 'Hosted notebook product (Code Ocean)',
    'Q14_5': 'Hosted notebook product (IBM Watson Studio)',
    'Q14_6': 'Hosted notebook product (Amazon Sagemaker Studio)',
    'Q14_7': 'Hosted notebook product (Amazon Sagemaker Studio Lab)',
    'Q14_8': 'Hosted notebook product (Amazon EMR Notebooks )',
    'Q14_9': 'Hosted notebook product (Google Cloud Vertex AI Workbench)',
    'Q14_10': 'Hosted notebook product (Hex Workspaces)',
    'Q14_11': 'Hosted notebook product (Noteable Notebooks)',
    'Q14_12': 'Hosted notebook product (Databricks Collaborative Notebooks)',
    'Q14_13': 'Hosted notebook product (Deepnote Notebooks)',
    'Q14_14': 'Hosted notebook product (Gradient Notebooks)',
    'Q14_14': 'Hosted notebook product (None)',
    'Q14_15': 'Hosted notebook product (Other)',
}

Q15 = {
    'Q15_1' : 'Data visualizations libraries (Matplotlib)',
    'Q15_2' : 'Data visualizations libraries (Seaborn)',
    'Q15_3' : 'Data visualizations libraries (Plotly/ Plotky Express)',
    'Q15_4' : 'Data visualizations libraries (Ggplot/ ggplot2)',
    'Q15_5' : 'Data visualizations libraries (Shiny)',
    'Q15_6' : 'Data visualizations libraries (D3 js)',
    'Q15_7' : 'Data visualizations libraries (Altair)',
    'Q15_8' : 'Data visualizations libraries (Bokeh)',
    'Q15_9' : 'Data visualizations libraries (Geoplotlib)',
    'Q15_10' : 'Data visualizations libraries (Leaflet/ Folium)',
    'Q15_11' : 'Data visualizations libraries (Pygal)',
    'Q15_12' : 'Data visualizations libraries (Dygraphs)',
    'Q15_13' : 'Data visualizations libraries (Highcharter)',
    'Q15_14' : 'Data visualizations libraries (None)',
    'Q15_15' : 'Data visualizations libraries (Other)',
}

Q17 = {
    'Q17_1' : 'ML frameworks (Scikit-learn)',
    'Q17_2' : 'ML frameworks (TensorFlow)',
    'Q17_3' : 'ML frameworks (Keras)',
    'Q17_4' : 'ML frameworks (Pytorch)',
    'Q17_5' : 'ML frameworks (Fast.ai)',
    'Q17_6' : 'ML frameworks (Xgboost)',
    'Q17_7' : 'ML frameworks (LightGBM)',
    'Q17_8' : 'ML frameworks (CatBoost)',
    'Q17_9' : 'ML frameworks (Caret)',
    'Q17_10' : 'ML frameworks (Tidymodels)',
    'Q17_11' : 'ML frameworks (JAX)',
    'Q17_12' : 'ML frameworks (Pytorch Lightning)',
    'Q17_13' : 'ML frameworks (Huggingface)',
    'Q17_14' : 'ML frameworks (None)',
    'Q17_15' : 'ML frameworks (Other)',
}

Q18 = {
    'Q18_1': 'ML Algorithms (Linear or Logistict Regression)',
    'Q18_2': 'ML Algorithms (Decison Trees or Random Forests)',
    'Q18_3': 'ML Algorithms (Gradient Boosting Machines)',
    'Q18_4': 'ML Algorithms (Bayesian Approaches)',
    'Q18_5': 'ML Algorithms (Evolutionary Approaches)',
    'Q18_6': 'ML Algorithms (Dense Neural Networks)',
    'Q18_7': 'ML Algorithms (Convolational Neural Networks)',
    'Q18_8': 'ML Algorithms (Generative Adverarial Networks)',
    'Q18_9': 'ML Algorithms (Recurrent Neural Networks)',
    'Q18_10': 'ML Algorithms (Transformer Networks)',
    'Q18_11': 'ML Algorithms (Autoencoder Networks)',
    'Q18_12': 'ML Algorithms (Graph Neural Networks)',
    'Q18_13': 'ML Algorithms (None)',
    'Q18_14': 'ML Algorithms (Other)',
}

Q19 = {
    'Q19_1' : 'Computer visions method (General tools)',
    'Q19_2' : 'Computer visions method (Image segmentation methods)',
    'Q19_3' : 'Computer visions method (Object detection methods)',
    'Q19_4' : 'Computer visions method (Image classification)',
    'Q19_5' : 'Computer visions method (Generative networks)',
    'Q19_6' : 'Computer visions method (Vision transformer networks)',
    'Q19_7' : 'Computer visions method (None)',
    'Q19_8' : 'Computer visions method (Other)',
}

Q20 = {
    'Q20_1' : 'NLP methods (Word embeddings/vectors)',
    'Q20_2' : 'NLP methods (Encoder-decoder models)',
    'Q20_3' : 'NLP methods (Contextualized embeddings)',
    'Q20_4' : 'NLP methods (Transformer language models)',
    'Q20_5' : 'NLP methods (None)',
    'Q20_6' : 'NLP methods (Other)',
}

Q21 = {
    'Q21_1' : 'Pretrain model services (TensorFlow Hub)',
    'Q21_2' : 'Pretrain model services (Pytorch Hub)',
    'Q21_3' : 'Pretrain model services (Huggingface models)',
    'Q21_4' : 'Pretrain model services (Timm)',
    'Q21_5' : 'Pretrain model services (Jumpstart)',
    'Q21_6' : 'Pretrain model services (ONNX models)',
    'Q21_7' : 'Pretrain model services (NVIDA NGC models)',
    'Q21_8' : 'Pretrain model services (Kaggle datasets)',
    'Q21_9' : 'Pretrain model services (None)',
    'Q21_10' : 'Pretrain model services (Other)',
}

Q22 = {
    'Q22_1' : 'ML repositories (Tfhub.dev)',
    'Q22_2' : 'ML repositories (Pytorch hub)',
    'Q22_3' : 'ML repositories (Huggingface models)',
    'Q22_4' : 'ML repositories (Timm)',
    'Q22_5' : 'ML repositories (Jumpstart)',
    'Q22_6' : 'ML repositories (ONNX models)',
    'Q22_7' : 'ML repositories (NVIDIA NGC models)',
    'Q22_8' : 'ML repositories (Kaggle datasets)',
    'Q22_9' : 'ML repositories (Other)',
}

Q28 = {
    'Q28_1' : 'Important role at work (Analyze and understand data)',
    'Q28_2' : 'Important role at work (Build/ run data infrastructure)',
    'Q28_3' : 'Important role (Build ML prototypes)',
    'Q28_4' : 'Important role ( Build/run ML services)',
    'Q28_5' : 'Important role (Improve existing ML models)',
    'Q28_6' : 'Important role (Research, advance ML models)',
    'Q28_7' : 'Important role (None)',
    'Q28_8' : 'Important role (Other)',
}

Q31 = {
    'Q31_1' : 'Cloud computing Platforms (Amazon Web Services)',
    'Q31_2' : 'Cloud computing Platforms (Microsoft Azure)',
    'Q31_3' : 'Cloud computing Platforms (Google Cloud Platform)',
    'Q31_4' : 'Cloud computing Platforms (IBM Cloud/ Red Hat)',
    'Q31_5' : 'Cloud computing Platforms (Oracle Cloud)',
    'Q31_6' : 'Cloud computing Platforms (SAP Cloud)',
    'Q31_7' : 'Cloud computing Platforms (VMware Cloud)',
    'Q31_8' : 'Cloud computing Platforms (Alibaba Cloud)',
    'Q31_9' : 'Cloud computing Platforms (Tencent Cloud)',
    'Q31_10' : 'Cloud computing Platforms (Huawei Cloud)',
    'Q31_11' : 'Cloud computing Platforms (None)',
    'Q31_12' : 'Cloud computing Platforms (Other)',
}

Q33 = {
    'Q33_1' : 'Cloud Computing products use (Amazon Elastic Compute Coud)',
    'Q33_2' : 'Cloud Computing products use (Microsoft Axure Virtual Machines)',
    'Q33_3' : 'Cloud Computing products use (Google Cloud Compute Engine)',
    'Q33_4' : 'Cloud Computing products use (None)',
    'Q33_5' : 'Cloud Computing products use (Other)',
}

Q34 = {
    'Q34_1' : 'Data storage products (Microsoft Azure Blob Storage)',
    'Q34_2' : 'Data storage products (Microsoft Azure Files)',
    'Q34_3' : 'Data storage products (Amazon Simple Storage Service)',
    'Q34_4' : 'Data storage products (Amazon Elastic File System)',
    'Q34_5' : 'Data storage products (Google CLoud Storage)',
    'Q34_6' : 'Data storage products (Gooogle Cloud Filestore)',
    'Q34_7' : 'Data storage products (None)',
    'Q34_8' : 'Data storage products (Other)',
}

Q35 = {
    'Q34_1' : 'Data products (MySQL)',
    'Q34_2' : 'Data products (PosgreSQL)',
    'Q34_3' : 'Data products (SQLite)',
    'Q34_4' : 'Data products (Oracle Database)',
    'Q34_5' : 'Data products (MongoDB)',
    'Q34_6' : 'Data products (SnowFlake)',
    'Q34_7' : 'Data products (IBM Db2)',
    'Q34_8' : 'Data products (Microsoft SQL Server)',
    'Q34_9' : 'Data products (Micorsoft Azure SQL Database)',
    'Q34_10' : 'Data products (Amazon Redshift)',
    'Q34_11' : 'Data products (Amazon RDS)',
    'Q34_12' : 'Data products (Amazon DynamoDB)',
    'Q34_13' : 'Data products (Google Cloud BigQuery)',
    'Q34_14' : 'Data products (Google Cloud SQL)',
    'Q34_15' : 'Data products (None)',
    'Q34_16' : 'Data products (Other)',
}

Q36 = {
    'Q36_1' : 'Business Intelligent tools (Amazon QuickSight)',
    'Q36_2' : 'Business Intelligent tools (Microsoft Power BI)',
    'Q36_3' : 'Business Intelligent tools (Google Data Studio)',
    'Q36_4' : 'Business Intelligent tools (Looker)',
    'Q36_5' : 'Business Intelligent tools (Tablue)',
    'Q36_6' : 'Business Intelligent tools (Qlik Sense)',
    'Q36_7' : 'Business Intelligent tools (Domo)',
    'Q36_8' : 'Business Intelligent tools (TIBCO Spotire)',
    'Q36_9' : 'Business Intelligent tools (ALteryx)',
    'Q36_10' : 'Business Intelligent tools (Silense)',
    'Q36_11' : 'Business Intelligent tools (SAP Ana;ytics Cloud)',
    'Q36_12' : 'Business Intelligent tools (Microsoft Azure Sysnapse)',
    'Q36_13' : 'Business Intelligent tools (Thoughtspot)',
    'Q36_14' : 'Business Intelligent tools (None)',
    'Q36_15' : 'Business Intelligent tools (Other)',
}

Q37 = {
    'Q37_1' : 'ML products (Amazon SageMaker)',
    'Q37_2' : 'ML products (Azure Machine Learning Studio)',
    'Q37_3' : 'ML products (Google CLoud Vertex AI)',
    'Q37_4' : 'ML products (DataRobot)',
    'Q37_5' : 'ML products (Databricks)',
    'Q37_6' : 'ML products (Dataiku)',
    'Q37_7' : 'ML products (Alteryx)',
    'Q37_8' : 'ML products (Rapidminer)',
    'Q37_9' : 'ML products (C3.ai)',
    'Q37_10' : 'ML products (Domino Data Lab)',
    'Q37_11' : 'ML products (H2O AI Cloud)',
    'Q37_12' : 'ML products (None)',
    'Q37_13' : 'ML products (Other)',

}

Q38 = {
    'Q38_1' : 'Automated ML tools (Google Cloud AutoML)',
    'Q38_2' : 'Automated ML tools (H2O Driverless AI)',
    'Q38_3' : 'Automated ML tools (Databricks AutoML)',
    'Q38_4' : 'Automated ML tools (DataRobot AutoML)',
    'Q38_5' : 'Automated ML tools (Amazon Sagemaker Autopilot)',
    'Q38_6' : 'Automated ML tools (Azure Automated Machine Learning)',
    'Q38_7' : 'Automated ML tools (None)',
    'Q38_8' : 'Automated ML tools (Other)',
}

Q39 = {
    'Q39_1' : 'ML products use (TensorFlow Extendted) ',
    'Q39_2' : 'ML products use (TorchServe) ',
    'Q39_3' : 'ML products use (ONNX Runtime) ',
    'Q39_4' : 'ML products use (Triton Inference Server) ',
    'Q39_5' : 'ML products use (Open VINO Model Server) ',
    'Q39_6' : 'ML products use (KServe) ',
    'Q39_7' : 'ML products use (BentoML) ',
    'Q39_8' : 'ML products use (Multi Model Server) ',
    'Q39_9' : 'ML products use (Seldon Core) ',
    'Q39_10' : 'ML products use (ML flow) ',
    'Q39_11' : 'ML products use (Other) ',
    'Q39_12' : 'ML products use (None) ',
}

Q40 = {
    'Q40_1' : 'ML monitors (Neptune.ai)',
    'Q40_2' : 'ML monitors (Weights & Biases)',
    'Q40_3' : 'ML monitors (Comet.ml)',
    'Q40_4' : 'ML monitors (TensorBoard)',
    'Q40_5' : 'ML monitors (Guild.ai)',
    'Q40_6' : 'ML monitors (ClearML)',
    'Q40_7' : 'ML monitors (MLflow)',
    'Q40_8' : 'ML monitors (Aporia)',
    'Q40_9' : 'ML monitors (Evidently AI)',
    'Q40_10' : 'ML monitors (Arize)',
    'Q40_11' : 'ML monitors (WhyLabs)',
    'Q40_12' : 'ML monitors (Fiddler)',
    'Q40_13' : 'ML monitors (DVC)',
    'Q40_14' : 'ML monitors (None)',
    'Q40_15' : 'ML monitors (Other)',
}

Q41 = {
    'Q41_1' : 'AI products (Google)',
    'Q41_2' : 'AI products (Microsoft)',
    'Q41_3' : 'AI products (IBM)',
    'Q41_4' : 'AI products (Amazon)',
    'Q41_5' : 'AI products (The LinkedIn)',
    'Q41_6' : 'AI products (Audit-AI)',
    'Q41_7' : 'AI products (Aequitas)',
    'Q41_8' : 'AI products (None)',
    'Q41_9' : 'AI products (Other)',
}

Q42 = {
    'Q42_1' : 'Hardware use (GPUs)',
    'Q42_2' : 'Hardware use (TPUs)',
    'Q42_3' : 'Hardware use (IPUs)',
    'Q42_4' : 'Hardware use (RDUs)',
    'Q42_5' : 'Hardware use (WSEs)',
    'Q42_6' : 'Hardware use (Trainium Chips)',
    'Q42_7' : 'Hardware use (Inferentia Chips)',
    'Q42_8' : 'Hardware use (None)',
    'Q42_9' : 'Hardware use (Other)',
}

Q44 = {
    'Q44_1' : 'Favorite media sources (Twitter)',
    'Q44_2' : 'Favorite media sources (Email newsletters)',
    'Q44_3' : 'Favorite media sources (Reddit)',
    'Q44_4' : 'Favorite media sources (Kaggle)',
    'Q44_5' : 'Favorite media sources (Course Forums)',
    'Q44_6' : 'Favorite media sources (YouTube)',
    'Q44_7' : 'Favorite media sources (Podcasts)',
    'Q44_8' : 'Favorite media sources (Blogs)',
    'Q44_9' : 'Favorite media sources (Journal Publications)',
    'Q44_10' : 'Favorite media sources (Slack)',
    'Q44_11' : 'Favorite media sources (None)',
    'Q44_12' : 'Favorite media sources (Other)',

}

In [68]:
ds_survey_df.rename(columns = one_choice_cols, inplace=True)
ds_survey_df.rename(columns = Q6, inplace=True)
ds_survey_df.rename(columns = Q7, inplace=True)
ds_survey_df.rename(columns = Q10, inplace=True)
ds_survey_df.rename(columns = Q12, inplace=True)
ds_survey_df.rename(columns = Q13, inplace=True)
ds_survey_df.rename(columns = Q14, inplace=True)
ds_survey_df.rename(columns = Q15, inplace=True)
ds_survey_df.rename(columns = Q17, inplace=True)
ds_survey_df.rename(columns = Q18, inplace=True)
ds_survey_df.rename(columns = Q19, inplace=True)
ds_survey_df.rename(columns = Q21, inplace=True)
ds_survey_df.rename(columns = Q22, inplace=True)
ds_survey_df.rename(columns = Q28, inplace=True)
ds_survey_df.rename(columns = Q31, inplace=True)
ds_survey_df.rename(columns = Q33, inplace=True)
ds_survey_df.rename(columns = Q34, inplace=True)
ds_survey_df.rename(columns = Q35, inplace=True)
ds_survey_df.rename(columns = Q36, inplace=True)
ds_survey_df.rename(columns = Q37, inplace=True)
ds_survey_df.rename(columns = Q38, inplace=True)
ds_survey_df.rename(columns = Q39, inplace=True)
ds_survey_df.rename(columns = Q40, inplace=True)
ds_survey_df.rename(columns = Q41, inplace=True)
ds_survey_df.rename(columns = Q42, inplace=True)
ds_survey_df.rename(columns = Q44, inplace=True)

In [69]:
ds_survey_df.head()

Duration (in seconds)    Age Gender Residential country Student?  \
0                   121  30-34    Man               India       No   
1                   462  30-34    Man             Algeria       No   
2                   293  18-21    Man               Egypt      Yes   
3                   851  55-59    Man              France       No   
4                   232  45-49    Man               India      Yes   

  DS Platform (Coursera) DS Platform (edX) DS Platform (Kaggle Learn Courses)  \
0                    NaN               NaN                                NaN   
1                    NaN               NaN                                NaN   
2               Coursera               edX                                NaN   
3               Coursera               NaN               Kaggle Learn Courses   
4                    NaN               NaN                                NaN   

  DS Platform (DataCamp) DS Platform (Fast.ai)  ...  \
0                    NaN                   NaN  ...   
1                    NaN                   NaN  ...   
2               DataCamp                   NaN  ...   
3                    NaN                   NaN  ...   
4                    NaN                   NaN  ...   

  Favorite media sources (Reddit)  Favorite media sources (Kaggle)  \
0                             NaN                              NaN   
1                             NaN                              NaN   
2                             NaN  Kaggle (notebooks, forums, etc)   
3                             NaN  Kaggle (notebooks, forums, etc)   
4                             NaN                              NaN   

              Favorite media sources (Course Forums)  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  Course Forums (forums.fast.ai, Coursera forums...   
4                                                NaN   

                    Favorite media sources (YouTube)  \
0                                                NaN   
1                                                NaN   
2  YouTube (Kaggle YouTube, Cloud AI Adventures, ...   
3                                                NaN   
4                                                NaN   

                   Favorite media sources (Podcasts)  \
0                                                NaN   
1                                                NaN   
2  Podcasts (Chai Time Data Science, O’Reilly Dat...   
3                                                NaN   
4                                                NaN   

                      Favorite media sources (Blogs)  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3  Blogs (Towards Data Science, Analytics Vidhya,...   
4  Blogs (Towards Data Science, Analytics Vidhya,...   

  Favorite media sources (Journal Publications)  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   

  Favorite media sources (Slack) Favorite media sources (None)  \
0                            NaN                           NaN   
1                            NaN                           NaN   
2                            NaN                           NaN   
3                            NaN                           NaN   
4                            NaN                           NaN   

  Favorite media sources (Other)  
0                            NaN  
1                            NaN  
2                            NaN  
3                            NaN  
4                            NaN  

[5 rows x 296 columns]

## Meaning of row

Each row is an response to the Kaggle state of Data Science Survey - 2022.

## Explore dataset


### 1. Missing values and duplicated rows

First, looking for any missing values in our dataframe

In [70]:
ds_survey_df.isnull().sum()


Duration (in seconds)                                0
Age                                                  0
Gender                                               0
Residential country                                  0
Student?                                             0
                                                 ...  
Favorite media sources (Blogs)                   16231
Favorite media sources (Journal Publications)    20193
Favorite media sources (Slack)                   22271
Favorite media sources (None)                    23997
Favorite media sources (Other)                   23162
Length: 296, dtype: int64

Now, let's see the missing values

In [71]:
missing_values = ds_survey_df[ds_survey_df.isnull().any(axis = 1)]
missing_values

Duration (in seconds)    Age Gender       Residential country Student?  \
0                       121  30-34    Man                     India       No   
1                       462  30-34    Man                   Algeria       No   
2                       293  18-21    Man                     Egypt      Yes   
3                       851  55-59    Man                    France       No   
4                       232  45-49    Man                     India      Yes   
...                     ...    ...    ...                       ...      ...   
23992                   331  22-24    Man  United States of America      Yes   
23993                   330  60-69    Man  United States of America      Yes   
23994                   860  25-29    Man                    Turkey       No   
23995                   597  35-39  Woman                    Israel       No   
23996                   303  18-21    Man                     India      Yes   

      DS Platform (Coursera) DS Platform (edX)  \
0                        NaN               NaN   
1                        NaN               NaN   
2                   Coursera               edX   
3                   Coursera               NaN   
4                        NaN               NaN   
...                      ...               ...   
23992                    NaN               NaN   
23993                    NaN               NaN   
23994                    NaN               NaN   
23995                    NaN               NaN   
23996                    NaN               NaN   

      DS Platform (Kaggle Learn Courses) DS Platform (DataCamp)  \
0                                    NaN                    NaN   
1                                    NaN                    NaN   
2                                    NaN               DataCamp   
3                   Kaggle Learn Courses                    NaN   
4                                    NaN                    NaN   
...                                  ...                    ...   
23992                                NaN                    NaN   
23993                                NaN                    NaN   
23994                                NaN               DataCamp   
23995               Kaggle Learn Courses                    NaN   
23996                                NaN                    NaN   

      DS Platform (Fast.ai)  ... Favorite media sources (Reddit)  \
0                       NaN  ...                             NaN   
1                       NaN  ...                             NaN   
2                       NaN  ...                             NaN   
3                       NaN  ...                             NaN   
4                       NaN  ...                             NaN   
...                     ...  ...                             ...   
23992                   NaN  ...                             NaN   
23993                   NaN  ...                             NaN   
23994                   NaN  ...                             NaN   
23995                   NaN  ...                             NaN   
23996                   NaN  ...                             NaN   

       Favorite media sources (Kaggle)  \
0                                  NaN   
1                                  NaN   
2      Kaggle (notebooks, forums, etc)   
3      Kaggle (notebooks, forums, etc)   
4                                  NaN   
...                                ...   
23992  Kaggle (notebooks, forums, etc)   
23993                              NaN   
23994  Kaggle (notebooks, forums, etc)   
23995                              NaN   
23996                              NaN   

                  Favorite media sources (Course Forums)  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3      Course Forums (forums.fast.ai, Coursera forums...   
4                

We can observe lots of missing values here, but those values are all catregorical values that are our survey options. Participants have the choice to select or not select these options, so we might disregard the missing values in this scenario.

Then, let's check if there is any duplicated rows.

In [72]:
print("Number of duplicated rows: ", np.count_nonzero(ds_survey_df.duplicated()))

Number of duplicated rows:  0


We see that there's no duplicated rows.

### 2. Check the invalid data

First, we will check the data types of columns in dataset

In [73]:
print(ds_survey_df.dtypes)

Duration (in seconds)                            object
Age                                              object
Gender                                           object
Residential country                              object
Student?                                         object
                                                  ...  
Favorite media sources (Blogs)                   object
Favorite media sources (Journal Publications)    object
Favorite media sources (Slack)                   object
Favorite media sources (None)                    object
Favorite media sources (Other)                   object
Length: 296, dtype: object


All the columns have data types is object and mainly have string values.
Remember we've made an observation that the Duration column should be of type float. Let's change its data type to float

In [74]:
ds_survey_df['Duration (in seconds)'] = ds_survey_df['Duration (in seconds)'].astype(np.float64)

### 3. How are values in numeric columns distributed?

In [75]:
ds_survey_df.describe()

Duration (in seconds)
count           2.399700e+04
mean            1.009010e+04
std             1.115403e+05
min             1.200000e+02
25%             2.640000e+02
50%             4.140000e+02
75%             7.150000e+02
max             2.533678e+06

The maximum duration takes to complete the survey is very strange. 

### 3. How are values in non-numeric columns distributed?

We wil find the value ratios for each questions, there is total 43 questions we need to work with.

First, we will consider colmuns with one choice.

In [76]:
def value_ratios(column):
    value_counts = column.value_counts()

    value_percent = ((value_counts / value_counts.sum()) * 100).round(1)
    value_percent_dict = value_percent.to_dict()

    return value_percent_dict

In [80]:
one_choice_cols_list = one_choice_cols.values()
one_choice_cols_df = ds_survey_df[one_choice_cols_list]

value_ratios_df = one_choice_cols_df.agg(value_ratios).to_frame()
value_ratios_df.columns = ['Value ratios']

value_ratios_df

Value ratios
Age                             {'18-21': 19.0, '25-29': 18.6, '22-24': 17.8, ...
Gender                          {'Man': 76.1, 'Woman': 22.0, 'Prefer not to sa...
Residential country             {'India': 36.6, 'United States of America': 12...
Student?                                                {'No': 50.2, 'Yes': 49.8}
Higher education?               {'Master’s degree': 39.1, 'Bachelor’s degree':...
Academic research publication?                          {'No': 57.6, 'Yes': 42.4}
Programming experience          {'1-3 years': 27.8, '< 1 years': 23.5, '3-5 ye...
ML experience                   {'Under 1 year': 36.3, '1-2 years': 18.7, 'I d...
Current industry                {'Computers/Technology': 25.5, 'Academics/Educ...
Company size                    {'0-49 employees': 23.5, '10,000 or more emplo...
Individual responible           {'20+': 25.3, '1-2': 20.2, '0': 16.2, '3-4': 1...
Incorporte ML methods           {'No (we do not use ML methods)': 22.2, 'We ha...
Current income                  {'$0-999': 13.7, '10,000-14,999': 6.1, '30,000...
Money spend for ML/Cloud        {'$0 ($USD)': 33.3, '$100-$999': 24.5, '$1000-...
Familiar Cloud Platform         {' Amazon Web Services (AWS) ': 28.8, ' Google...
TPU experience                  {'Never': 55.3, '2-5 times': 20.9, 'Once': 13....

In [ ]:
#TODO: add the distribution of values in multiple choice columns